packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -225,30 +202,70 @@ class DualVariableGadget : public libsnark::dual_variable_gadget<FieldT>
     }
 };
 
-// Helper function that contains the history of all the values of a variable
-class DynamicVariableGadget : public GadgetT
+class ToBitsGadget : public libsnark::dual_variable_gadget<FieldT>
 {
   public:
-    std::vector<VariableT> variables;
-    bool allowGeneratingWitness;
 
-    DynamicVariableGadget( //
+    ToBitsGadget( //
       ProtoboardT &pb,
+      const VariableT &value,
+      const size_t width,
       const std::string &prefix)
-        : GadgetT(pb, prefix)
+        : libsnark::dual_variable_gadget<FieldT>(pb, value, width, prefix)
+    {
+
+    }
+
+    void generate_r1cs_witness()
     {
-        add(make_variable(pb, FMT(prefix, ".initialValue")));
-        allowGeneratingWitness = true;
+        generate_r1cs_witness_from_packed();
     }
 
+    void generate_r1cs_constraints()
+    {
+        libsnark::dual_variable_gadget<FieldT>::generate_r1cs_constraints(true);
+    }
+};
+
+typedef ToBitsGadget RangeCheckGadget;
+
+class FromBitsGadget : public libsnark::dual_variable_gadget<FieldT>
+{
+  public:
+
+    FromBitsGadget( //
+      ProtoboardT &pb,
+      const VariableArrayT &bits,
+      const std::string &prefix)
+        : libsnark::dual_variable_gadget<FieldT>(pb, bits, prefix)
+    {
+
+    }
+
+    void generate_r1cs_witness()
+    {
+        generate_r1cs_witness_from_bits();
+    }
+
+    void generate_r1cs_constraints(bool enforce = true)
```

Comment:
Can't hurt to have it I think. The class is a simple wrapper around `dual_variable_gadget` which also exposes it and there are some cases it may not be necessary which could save quite a bit of constraints because a lot of times there are many bits.